### ISSUE TITLE
feat(model): wire Engram packed-row retrieval into the V4.1 forward

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#13007]
# feat(model): wire Engram packed-row retrieval into the V4.1 forward

# feat(model): wire Engram packed-row retrieval into the V4.1 forward

``routing
lane: model
paths: ["internal/model/v41_forward.go", "internal/model/v41_cache.go"]
``

**Repo:** `anthony-chaudhary/fak` (public core) -- **Ref** #12640 -- **Related (closed)** #12936, #12904

## Parent context

#12640; related (closed) #12936, #12904.

## Why now

The V4.1 text forward is documented as omitting the Engram stage, and the row cache is not wired into `v41Layer`, so the assembled pass is not the full V4.1 text path.

## Working spine

Wire `V41EngramRowCache.Row` (via `GatherV41EngramRows`) into `v41Layer` so each layer's Engram output is retrieved, dequantized, and mixed per the reference schedule, failing closed when the source is absent.

## Out of scope

Depends on issue 15 (disk-backed row source) for real checkpoint IO but does not block on it: the in-memory `V41EngramRowSource` suffices to land this leaf.

## Acceptance gate

`go -C ..\fak test ./internal/model -run TestV41EngramRetrievalForward -count=1`; full-checkpoint numeric parity remains `[SW-VERIFIED]` (reduced-model oracle) until a real on-device run.

## Closure binding

The resolving commit cites the tracking issue and carries `(fak model)`.

## Problem

The V4.1 text forward is documented as omitting the Engram stage. `internal/model/v41_forward.go:13-15` states:

> no Engram packed-row retrieval (that stage's rows are 264-byte packed FP8 streams a weight-free in-memory forward cannot materialize; it is an explicit, documented omission)

The supporting machinery already exists and is complete in isolation:

- `internal/model/v41_engram.go:133` `Hash` " token -> `[token][layer][ngram-head]` row-ID lookup.
- `internal/model/v41_engram.go:192` `GatherV41EngramRows` " resolves flattened hash output through one bounded cache per layer.
- `internal/model/v41_cache.go:161-167` `V41EngramRowSource` " backing read seam (in-memory/synthetic injectable).
- `internal/model/v41_cache.go:223` `V41EngramRowCache.Row` " cache hit/read-on-miss with bounded eager prefetch.

**Gap:** the row cache is not wired into `v41Layer` (`v41_forward.go:377`); the forward path never retrieves Engram rows. Even when a `V41EngramRowSource` is supplied, the assembled pass is not the full V4.1 text path.

## Deliverable

1. Wire `V41EngramRowCache.Row` (via `GatherV41EngramRows`) into `v41Layer` so each layer's Engram output is retrieved, dequantized, and mixed per the reference schedule.
2. Fail **closed** when the row source/Engram stage is absent: return an `ErrV41NativeUnsupported`-class error. No fall-through to a zero/no-op embedding, no silent mock.
3. Keep the stage admission-gated in `v41ForwardAdmitted`/`forwardV41` (existing fail-closed pattern).

## Witness

``
go -C ..\fak test ./internal/model -run TestV41EngramRetrievalForward -count=1
``

Full-checkpoint numeric parity remains `[SW-VERIFIED]` (reduced-model oracle) until a real on-device run; do not mark `[HW-WITNESSED]`.

## Negative test

Absent row source must return an `ErrV41NativeUnsupported`-class failure (`errors.Is` true), never a silently-zero Engram contribution.

## Acceptance criteria (definition of done)
- [ ] Engram rows retrieved + dequantized + mixed inside `v41Layer` per reference schedule.
- [ ] Missing row source errors fail-closed (`ErrV41NativeUnsupported`-class); no zero-embedding fall-through.
- [ ] `TestV41EngramRetrievalForward` green; existing `v41_forward_test.go` oracle unchanged.
- [ ] `gofmt`/`go vet` clean; `go run ./cmd/fak-boundary check --all` exit 0.

## Likely files

- `internal/model/v41_forward.go`
- `internal/model/v41_cache.go`

## Boundary

Public core (`fak`, `internal/model`) " no private imports, no `fak-private` dependency.

## Dependency

Depends on issue 15 (disk-backed row source) for real checkpoint IO, but **does not block** on it: the in-memory `V41EngramRowSource` suffices to land this leaf.

**Comment by anthony-chaudhary** [comment]
**Bounded leaf landed: the forward now fails closed on a declared-but-unexecuted Engram layer.**

Landed (public ak main): dd2039a4ebff7ca4b617126ffbf6f2bdaa90f783 (ff 60ec00e83..dd2039a4e).

### What was missing

41ForwardAdmitted (internal/model/v41_forward.go) never checked the config's declared Engram layers. The reduced V4.1 text assembly deliberately omits the Engram stage, so a config declaring an Engram layer **within** the decoder stack ran the forward and emitted non-Engram logits as if that layer did not exist — a silent omission. The published 40-layer artifact declares engram_layer_ids: [1, 14], both in range.

### Delivered

internal/model/v41_forward.go (+35/-2) and internal/model/v41_forward_test.go (+27):

- New 41StageEngram stage constant.
- New 41EngramForwardAdmitted(cfg): refuses at admission with a typed *V41ForwardError wrapping ErrV41ForwardStage that names the offending layer, whenever a declared Engram layer lies in [0, NumLayers). Declared Engram layers outside the stack stay admitted, so the reduced oracle fixture (NumLayers=1, EngramLayerIDs [1,14]) keeps running.
- TestV41ForwardEngramDeclaredFailsClosed: in-range declaration refuses at both admission and Forward; the reduced out-of-range fixture stays admitted (no regression).

This lands the "**failing closed when the source is absent**" clause of this issue. The remaining integration clause — actually executing the Engram stage and mixing its retrieved rows per the reference schedule — is unchanged and still open.

### Witness (RED / GREEN)

`	ext
go test ./internal/model -run TestV41ForwardEngramDeclaredFailsClosed -count=1
  RED   (pre-fix):  in-range Engram admission error = <nil>, want ErrV41ForwardStage  => FAIL
  GREEN (post-fix): PASS
go test ./internal/model/... -count=1                                  => ok 164.462s
go vet ./internal/model                                                => exit 0
gofmt -l internal/model/v41_forward{,_test}.go                         => clean
go run ./cmd/fak-boundary check --staged --fak-dir <worktree>          => PASS (0 violations, 5 gates)
`

Rebased onto origin/main@60ec00e83 and re-verified before landing; never force-pushed.

Ref #12640. No native V4.1 generation and no physical parity is claimed; vision/audio/DSpark remain out of scope. V4 Flash 0731 (#12635) control work was not touched.

**Comment by anthony-chaudhary** [comment]
**Landed: the Engram stage now executes in the V4.1 forward — retrieved, dequantized, and mixed per the reference schedule.**

Landed (public `fak` main): `6780607fd22d` (`internal/model/v41_forward.go`, `internal/model/v41_forward_engram.go` (new), `internal/model/v41_forward_test.go`; 900 insertions), integrated onto `origin/main@94c847015`.

### What was missing

`v41EngramForwardAdmitted` refused every in-range declared Engram layer (the #13007 fail-closed seam). The retrieval leaves — `V41EngramHashState.Hash`, `GatherV41EngramRows`, `V41EngramRowCache.Row`, the 264-byte E4M3+E8M0 packed-row layout — existed and were witnessed in isolation, but nothing executed them on the forward path, so a config declaring an Engram layer either refused or (before the earlier seam) emitted non-Engram logits.

### Delivered

- `v41_forward_engram.go` (new): `v41EngramStage` (hash state + one bounded row cache per declared Engram layer), `wireV41Engram`, `dequantV41EngramRow` (E4M3 + E8M0, reject the E8M0 0xff / E4M3 NaN bytes, bf16-round each value), `v41EngramInject`, and a `v41BF16` round-half-to-even helper.
- `v41Layer` now calls `v41EngramInject` at the **START of the layer, before attention** — the reference position (`antirez/ds4@bd66c402`, `ds41_graph_before_attention` + `metal/dsv41.metal` `kernel_dsv41_engram_add`): hash the token history to 24 row ids, retrieve/dequantize, project through `engram_kv.weight` into `(N_HC+1)` streams, then per HC stream compute the RMS-normalized learned-key dot, `gate = sigmoid(copysign(sqrt(max(|dot|,1e-6)), dot))`, and add `bf16(gate*bf16(value))` into the residual.
- `v41EngramForwardAdmitted` now admits an in-range Engram layer only when the model carries a wired row source and the three mixing tensors; an absent source returns an error wrapping `ErrV41NativeUnsupported` (the issue's negative-test class). Out-of-range declarations stay admitted (no regression on the reduced oracle fixture).

### Witness (SW-VERIFIED, RED / GREEN)

```
go test ./internal/model -run TestV41EngramRetrievalForward -count=1
  RED   (inject stubbed): logits[0]=0.3852725 want -0.09182708 |delta|=4.771e-01 => FAIL
  GREEN (wired):          PASS
go test ./internal/model -count=1 => ok 171.064s
go vet ./internal/model           => exit 0
gofmt -l (3 files)                => clean
```

The witness is an independent scalar transcription of the projection+gate+mix schedule (it does not call the production inject/dequant/bf16 helpers); it shares only the retrieval primitives the stage wires. A `cross-validator` pass returned PASS-WITH-CONCERNS (doc-precision only, fixed).

### Honesty boundary

Full-checkpoint numeric parity remains **[SW-VERIFIED]** against a synthetic in-memory `V41EngramRowSource`; **no hardware witness** is claimed, and this leaf does not touch the compute kernels. The reduced model collapses the reference's four identical stand-in HC streams into one vector — documented in the source — so this is the injection schedule, not official-checkpoint generation.

### Landing note

`fak commit`'s prospective gate classifies any `internal/model` path as GPU-relevant and requires the Linux-only Strix controller authority plus a reachable `strix1` appliance. On this Windows host the authority is structurally unavailable, so the commit was landed through a **clean-stamped WSL build** (the sanctioned `UNSUPPORTED_VALIDATOR_BINARY` recovery). Build, vet, gofmt, and the full affected-package tests all pass; only the physical-appliance phase was unsatisfied, which is why the landing used `--no-build-check` with the evidence above recorded.

Ref #12640. Related (closed) #12936, #12904; the dependency #12993's disk-backed `internal/ggufload/v41_engram_q2k_rows.go` landed independently and was not required for this in-memory leaf.

**Comment by anthony-chaudhary** [comment]
Resolved by 6780607fd22d (landed on origin/main as 94c847015). The Engram stage now executes in the reduced V4.1 forward — hash -> gather -> dequantize (E4M3/E8M0, bf16) -> engram_kv projection -> RMS-normalized learned-key gate -> residual mix at the layer start, per antirez/ds4@bd66c402; absent row source fails closed with ErrV41NativeUnsupported. Acceptance witness TestV41EngramRetrievalForward + TestV41EngramNoSourceFailsClosed green; full internal/model package green. Numeric parity remains [SW-VERIFIED] (synthetic in-memory row source).
